File: Day1b/Day1b.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

// Captcha as defined in the puzzle:
// Now, instead of considering the next digit, it wants you to consider the
// digit halfway around the circular list. That is, if your list contains 10
// items, only include a digit in your sum if the digit 10/2 = 5 steps forward
// matches it. Fortunately, your list has an even number of elements.
func Captcha(seq string) (int, error) {
	var sum int

	seql := strings.Split(seq, "")
	seqlen := len(seql)
	step := int(seqlen / 2)

	for i, s := range seql {
		if s == seql[(i+step)%seqlen] {
			v, err := strconv.Atoi(s)
			if err != nil {
				return sum, err
			}
			sum += v
		}
	}
	return sum, nil
}

func usage() {
	fmt.Fprintf(os.Stderr, "usage: %s [sequence]\n", os.Args[0])
	flag.PrintDefaults()
	os.Exit(2)
}

func main() {
	flag.Usage = usage
	flag.Parse()

	args := flag.Args()
	if len(args) < 1 {
		fmt.Println("Sequence missing.")
		os.Exit(1)
	}
	// No sanity checking on input here either.
	retval, err := Captcha(args[0])
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(retval)
}

File: Day1b/Day1b.py
#!/usr/bin/python3

import sys


# Now, instead of considering the next digit, it wants you to consider the
# digit halfway around the circular list. That is, if your list contains 10
# items, only include a digit in your sum if the digit 10/2 = 5 steps forward
# matches it. Fortunately, your list has an even number of elements.
def captcha(seq):
    # Split sequence into individual digits.
    seql = [int(x) for x in seq]

    # Perform the operation on the list
    sum = 0
    seqlen = len(seq)
    step = int(seqlen/2)
    for i in range(len(seq)):
        v = seql[i]
        if v == seql[(i+step) % seqlen]:
            sum += v
    return sum

# Unit tests for captcha.
tt = {'1212': 6, '1221': 0, '123425': 4, '123123': 12, '12131415': 4}
for k, v in tt.items():
    result = captcha(k)
    if result != v:
        print("FAIL: input ", k, ": expected ", v, ", got ", result, sep="")

# The input is not checked for sanity, just existence.
seq = sys.stdin.readlines()
if len(seq) == 0:
    print("Sequence missing!")
    sys.exit(1)

# Only one line, and I don't want the newline character.
print(captcha(seq[0][:-1]))
